src/ape_cache/query.py

```diff
@@ -0,0 +1,472 @@
+            select([func.count()])
+            .select_from(Transactions)
+            .where(Transactions.block_hash == query.block_id)
+        )
+
+    @_estimate_query_clause.register
+    def _contract_events_estimate_query_clause(self, query: ContractEventQuery) -> Select:
+        return (
+            select([func.count()])
+            .select_from(ContractEvents)
+            .where(ContractEvents.block_number >= query.start_block)
+            .where(ContractEvents.block_number <= query.stop_block)
+            .where(ContractEvents.block_number % query.step == 0)
+        )
+
+    @singledispatchmethod
+    def _compute_estimate(self, query: QueryType, result: CursorResult) -> Optional[int]:
+        """
+        A singledispatchemethod that computes the time a query
+        will take to perform from the caching database
+        """
+
+        return None  # can't handle this query
+
+    @_compute_estimate.register
+    def _compute_estimate_block_query(
+        self,
+        query: BlockQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.scalar() == (1 + query.stop_block - query.start_block) // query.step:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        # TODO: Allow partial queries
+        return None
+
+    @_compute_estimate.register
+    def _compute_estimate_block_transaction_query(
+        self,
+        query: BlockTransactionQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.scalar() > 0:  # type: ignore
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        return None
+
+    @_compute_estimate.register
+    def _compute_estimate_contract_events_query(
+        self,
+        query: ContractEventQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.scalar() == (query.stop_block - query.start_block) // query.step:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        # TODO: Allow partial queries
+        return None
+
+    def estimate_query(self, query: QueryType) -> Optional[int]:
+        """
+        Method called by the client to return a query time estimate.
+
+        Args:
+            query (QueryType): Choice of query type to perform a
+                check of the number of rows that match the clause.
+
+        Returns:
+            Optional[int]
+        """
+
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self._estimate_query_clause(query))
+                if not result:
+                    return None
+
+                return self._compute_estimate(query, result)
+
+        except QueryEngineError as err:
+            logger.warning(f"Cannot perform query on cache database: {err}")
+            return None
```

Comment:
Note that the reason we instead return `None` here instead of failing is that we want a failure of the query to bypass the query logic so that the estimation phase does not fail in `QueryManager`

Reply:
[21465f7](https://github.com/ApeWorX/ape/pull/680/commits/21465f70f97a56ad08ed00e35500fc40864eef1d)